### PR TITLE
Interactive TUX

### PR DESCRIPTION
**Pull request by mxsdev**
Solves #100 

<p align="center">

</p>

### CODE REVIEW

**Comment by mxsdev** [comment]
@sourishkrout Oh yeah, sorry, I misunderstood the original ticket. Fixed in [4c33c94](https://github.com/stateful/runme/pull/103/commits/4c33c94bea59a33afd1021438b938bbf0974f222) and I added an `--exit` flag for the case where you want to run a command one-off.

**Comment by sourishkrout** [comment]
Hey @silas, you have a sec to check out whether or not the UX meets your expectations?

Outside of that, @mxsdev, one thing I noticed is that the list entries can exceed your terminal's height with a fully loaded README.md such as https://github.com/stateful/vscode-runme/blob/main/examples/README.md. We could track this issue in a separate ticket, though.

**Comment by silas** [comment]
I check tried the branch locally and it works great!

The one comment I'd add is that if the section fails it drops you out of the tui and prints a kinda ugly error message.

```sh
runme

> echo-ok sh
    echo ok
  echo-fail
    echo fail (...)

ok
runme

  echo-ok
    echo ok
> echo-fail sh
    echo fail (...)

fail
failed to run command "set -e -o pipefail;echo fail;false;\n": exit status
```

For the following readme

~~~markdown
```sh
echo ok
```

```sh
echo fail
false
```
~~~

Might be worth just printing an error message with the exist status code and sending you back into the selection menu.

Might also be nice to add a newline after the run, just to create some space between the run output and the `runme` title.

**Comment by mxsdev** [comment]
Added scrolling (number of shown entries configurable with `--entries) and recovery from error. Also made the spacing nicer between the command's stdout and the runme tui. (and I fixed the linting errors...[for real this time](https://github.com/mxsdev/runme/actions/runs/3960604421) 😅 )

<p align="center">

</p>

I didn't give any scroll indicators, so the ui is a little minimal in this respect. Definitely let me know if you think we should add more to the ui.

@silas The ugly error message should be tracked in a separate issue, since this is the same behavior as in executing `runme run <cmd>`.

**Comment by silas** [comment]
@mxsdev looks great! Also nice spacing change, love the polish.

I created issue #106 about improving the error message.

**Comment by sourishkrout** [comment]
Great stuff @mxsdev! Good call on spinning out #106. @adambabik could you do a review please?

File: internal/cmd/root.go
Comment (adambabik) [inline internal/cmd/root.go:19]: Having `tuiCmd` flags here while they belong to `tuiCmd` as well as interleaving `tuiCmd` with the root command might be improved a bit.

In `root.go`, I would add these lines:

```go
tuiCmd := tuiCmd()
// tuiCmd is also a default command
cmd.RunE = tuiCmd.RunE
// remember to copy flags
tuiCmd.Flags().VisitAll(func(f *pflag.Flag) {
	pflags.AddFlag(f)
})
```

And:

```go
cmd.AddCommand(tuiCmd)
```

In this way, we keep `tuiCmd` totally unaware that it is used in any different way than other commands.

**Review by adambabik — COMMENTED** [review]
It's really good stuff 👍 It works very well!

I have two major comments about the code:
- `tuiModel` can be improved (removal of unnecessary pointers) so that it's more aligned with how bubbletea encourages to write programs.
- `tuiCmd()` can look like any other command and you can avoid interleaving some of its fragments with the root command.

I am happy to talk in more detail on Discord if you wish.

Thanks for the PR!

File: internal/cmd/tui.go
Comment (mxsdev) [reply internal/cmd/tui.go]: The reason I had all those pointers initially was because I didn't realize `prog.Run()` returns the model, haha. Thanks for pointing it out!

File: internal/cmd/root.go
Comment (mxsdev) [reply internal/cmd/root.go]: The flag copying code errors because it re-defines the flags set by `setDefaultFlags`. In the most recent PR I used the following code:

```go
	tuiCmd.Flags().VisitAll(func(f *pflag.Flag) {
		if flag := cmd.Flags().Lookup(f.Name); flag == nil {
			cmd.Flags().AddFlag(f)
		}
	})
```

The alternative would be to never run `setDefaultFlags` on the root command and let them cascade from the tui command. I felt that it was better to be explicit than implicit here, even if at the expense of verbosity. Let me know if you want me to do it the other way!

**Comment by mxsdev** [comment]
@adambabik Should all be implemented as you said, let me know if I missed anything!

File: internal/cmd/run.go
Comment (adambabik) [inline internal/cmd/run.go:58]: To align the interface with `newExecutable()` you can change it to:

```suggestion
func runBlock(cmd *cobra.Command, block *document.CodeBlock, opts *runCmdOpts) error {
```

File: internal/cmd/tui.go
Comment (adambabik) [inline internal/cmd/tui.go:228]: You can use a technique called the early return to avoid unnecessary nesting:

```go
if result.run != nil {
    break
}

if err := runBlockCmd(result.run, cmd, nil); err != nil {
   ...
}
```

**Review by adambabik — APPROVED** [review]
Thank you! I added a few nitpick comments, but it's good to be merged :)

**Comment by adambabik** [comment]
@mxsdev I have one more ask - can you resolve conflicts, please?

**Comment by mxsdev** [comment]
Resolved conflicts, but I realized there is an error when you have a README.md file with no scripts. 

I think in this case we should probably just show a message like "No scripts to run in README.md!" And then close the program. Do you want to open a separate issue for this? Or just solve it here?

**Comment by sourishkrout** [comment]
> Resolved conflicts, but I realized there is an error when you have a README.md file with no scripts.
> 
> I think in this case we should probably just show a message like "No scripts to run in README.md!" And then close the program. Do you want to open a separate issue for this? Or just solve it here?

I'd say let's get this merged and work on the issue above in a separate PR. We can hold off cutting a stable release until that has landed.

**Comment by amac4015** [comment]
iam have to say maybe a little more i dont know trainning or someting is in order since i dont really have tpo muck knowalge on whats the scope of the progect

**Comment by amac4015** [comment]
or not its cool good work guys

**Pull request merged**
